Subscribe the newsletter if email_newsletter is changed

takmar opened the pull request:
<!--
     For Work In Progress Pull Requests, please use the Draft PR feature,
     see https://github.blog/2019-02-14-introducing-draft-pull-requests/ for further details.

     For a timely review/response, please avoid force-pushing additional
     commits if your PR already received reviews or comments.

     Before submitting a Pull Request, please ensure you've done the following:
     - 📖 Read the Forem Contributing Guide: https://github.com/forem/forem/blob/main/CONTRIBUTING.md#create-a-pull-request.
     - 📖 Read the Forem Code of Conduct: https://github.com/forem/forem/blob/main/CODE_OF_CONDUCT.md.
     - 👷‍♀️ Create small PRs. In most cases this will be possible.
     - ✅ Provide tests for your changes.
     - 📝 Use descriptive commit messages.
     - 📗 Update any related documentation and include any relevant screenshots.

     NOTE: Pull Requests from forked repositories will need to be reviewed by
     a Forem Team member before any CI builds will run. Once your PR is approved
     with a `/ci` reply to the PR, it will be allowed to run subsequent builds without
     manual approval.
-->

## What type of PR is this? (check all applicable)

- [ ] Refactor
- [ ] Feature
- [x] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description
In `Users::NotificationSetting` model, `Users::SubscribeToMailchimpNewsletterWorker` is executed **ONLY** the case `email_newsletter` is true. So a user can not unsubscribe from the newsletter by updating `email_newsletter` to false.

To fix this, I changed `subscribe_to_mailchimp_newsletter` method to check `saved_changes.key?(:email_newsletter)` instead of `email_newsletter`.

And I also make it to check `Settings::General.mailchimp_api_key.blank?` the same as `User` model.
https://github.com/forem/forem/blob/e50db312fb5181c674a80b0c0e86f06752be8afc/app/models/user.rb#L509-L511

## Related Tickets & Documents

## QA Instructions, Screenshots, Recordings

_Please replace this line with instructions on how to test your changes, a note
on the devices and browsers this has been tested on, as well as any relevant
images for UI changes._

### UI accessibility concerns?

_If your PR includes UI changes, please replace this line with details on how
accessibility is impacted and tested. For more info, check out the
[Forem Accessibility Docs](https://docs.forem.com/frontend/accessibility)._

## Added/updated tests?

- [x] Yes
- [ ] No, and this is why: _please replace this line with details on why tests
      have not been included_
- [ ] I need help with writing tests

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

- [ ] I've updated the [Developer Docs](https://docs.forem.com) and/or
      [Admin Guide](https://admin.forem.com/), or
      [Storybook](https://storybook.forem.com/) (for Crayons components)
- [ ] I've updated the README or added inline documentation
- [ ] I will share this change in a [Changelog](https://forem.dev/t/changelog)
      or in a [forem.dev](http://forem.dev) post
- [ ] I will share this change internally with the appropriate teams
- [ ] I'm not sure how best to communicate this change and need help
- [ ] This change does not need to be communicated, and this is why not: _please
      replace this line with details on why this change doesn't need to be
      shared_

## [optional] Are there any post deployment tasks we need to perform?

## [optional] What gif best describes this PR or how it makes you feel?

![alt_text](gif_link)

takmar commented on spec/models/users/notification_setting_spec.rb:
```diff
@@ -0,0 +1,44 @@
+
+      it "enqueues SubscribeToMailchimpNewsletterWorker when updating email_newsletter to false" do
+        notification_setting.update(email_newsletter: true)
+        sidekiq_assert_enqueued_jobs(1, only: Users::SubscribeToMailchimpNewsletterWorker) do
```
I used not `sidekiq_assert_enqueued_with` but `sidekiq_assert_enqueued_jobs` for checking the jobs count. Because L24 pushes a `Users::SubscribeToMailchimpNewsletterWorker` job as well.

msarit commented on app/models/users/notification_setting.rb:
```diff
@@ -13,7 +13,8 @@ class NotificationSetting < ApplicationRecord
 
     def subscribe_to_mailchimp_newsletter
-      return unless email_newsletter
+      return if Settings::General.mailchimp_api_key.blank?
```
@takmar This is an amazing catch Takmar, thank you so much! 🙏🏾

msarit commented on spec/models/users/notification_setting_spec.rb:
```diff
@@ -0,0 +1,44 @@
+  let!(:user) { create(:user) }
+  let(:notification_setting) { described_class.find_by(user_id: user.id) }
+
+  describe "validations" do
```
I'm tagging @citizen428 and @rhymes to get their take on this, @takmar, but I **believe we can remove this entire `describe` block**, and the corresponding solitary validation in app/models/users/notification_setting.rb:7 (line 7).

When running the spec, I get the following warning from `shoulda-matchers` gem:
```
Warning from shoulda-matchers:

You are using `validate_inclusion_of` to assert that a boolean column
allows boolean values and disallows non-boolean ones. Be aware that it
is not possible to fully test this, as boolean columns will
automatically convert non-boolean values to boolean ones. Hence, you
should consider removing this test.
```

citizen428 replied:
Alas, this column doesn't have a non-null constraint so we would be able to accidentally unset this without the validation. Since this is tangential to this PR I think we should leave it in and clean up all the boolean validations at the same time, adding constraints where necessary.

citizen428 replied:
Yes, we can remove this, see http://www.chrisrolle.com/en/blog/boolean-attribute-validation.

takmar replied:
I removed the code related to validation. Thank you!
https://github.com/forem/forem/pull/14226/commits/93095a325e30e29fdd1f8369644711f956a096c0

msarit reviewed (changes requested):
Great job, @takmar! 🚀 

Just a few small changes requested.

citizen428 commented on spec/models/users/notification_setting_spec.rb:
```diff
@@ -0,0 +1,44 @@
+
+RSpec.describe Users::NotificationSetting, type: :model do
+  let!(:user) { create(:user) }
+  let(:notification_setting) { described_class.find_by(user_id: user.id) }
```
**Suggestion (non-blocking):** This also works and may be a bit more straightforward:

```suggestion
  let(:notification_setting) { user.notification_setting.reload }
```

takmar commented:
I've noticed that the email column is nullable, so I also added the line to check an email existence to avoid enqueuing a job to the newsletter without an email.
https://github.com/forem/forem/pull/14226/commits/940601c51794d143dd897ee4a0fa083d320bb1f0

rhymes commented:
@takmar thanks for the changes! Looks like there are two legit failures in the tests here https://app.travis-ci.com/github/forem/forem/jobs/524379134#L913-L914

```
rspec ./spec/requests/user/user_notification_settings_spec.rb:40 # UserNotificationSettings PUT /update/:id returns error message if settings can't be saved

rspec ./spec/requests/user/user_notification_settings_spec.rb:85 # UserNotificationSettings PATCH /onboarding_notifications_checkbox_update returns 422 status and errors if errors occur
```

Can you take a look?

takmar commented:
@rhymes 
Thank you for the information!

The reason for both of the test failures is that the controller subjected to testing raised the ActiveRecord::NotNullViolation error when those tests set nil to the non-nullable email_digest_periodic column. This is because of the removal of the validation for email_digest_periodic.

Though the failure tests expected the controller to set an error message in case of validation failure, there are no longer validations on the model. So we can not test that case anymore.

I established the following ways to manage the failures:
- Remove simply those tests.
- Add boolean column validations to the model in another way that does not show a warning.

Though I removed the validation showing the warning, [Rails Guides introduce another way](https://guides.rubyonrails.org/active_record_validations.html#inclusion) to validate a boolean column.

```
validates :boolean_field_name, exclusion: [nil]
```

When I tried to implement a validation this way and added the test for it with the validate_exclusion_of matcher, the warnings no longer appeared. This will prevent unexpectedly raising the error and set an error message to a model instance, so I think this could be a better way to deal with the failures.

What do you think about these? Or do you have any other ideas?

citizen428 commented:
> validates :boolean_field_name, exclusion: [nil]

@takmar Wouldn't `presence: true` also work and be a bit more straightforward?

takmar commented:
@citizen428 
Though it is certainly easy to understand, In the case a boolean column value is `false`, presence validation always fails. Because the `PresenceValidator` calls the `blank?` method for validation. So we can't use presence validation for a boolean column.

https://github.com/rails/rails/blob/main/activemodel/lib/active_model/validations/presence.rb#L5-L8

citizen428 commented:
Yup @takmar, my bad, I didn't properly take the context of our discussion (booleans) into account 🤦‍♂️ 😃 

I'm personally of the opinion that boolean columns should always be non-nullable dealing with a tri-state is almost never what one wants.

takmar commented:
> dealing with a tri-state is almost never what one wants

@citizen428 Let me check if I understood correctly. Does this mean it's better to add a validation to check if nil for a boolean column?

citizen428 commented:
I just meant that boolean columns should pretty much always have a non-null constraint in the DB. Whether or not we put a validation in front of that depends a lot on where/how a value is used, i.e. how important it is to communicate this back to the user in a friendlier way than an exception.

citizen428 commented:
`It seems we have legitimate spec failures right now. On a similar PR we recently made the call to keep the boolean validation in until we have a more comprehensive strategy across the app. So if you want to re-add it that's fine.

takmar commented:
I'm sorry for my late response. I pushed another commit to fix the failures in the tests. I ended up adding validations in the way [I described before](https://github.com/forem/forem/pull/14226#issuecomment-882000531
).

> On a similar PR we recently made the call to keep the boolean validation in until we have a more comprehensive strategy across the app. So if you want to re-add it that's fine.

Though I implemented the validations for now, feel free to remove them when you decided on the strategy.

aitchiss commented:
@msarit @citizen428 just a quick ping as the specs are passing now 🙂

citizen428 commented on app/models/users/notification_setting.rb:
```diff
@@ -4,7 +4,21 @@ class NotificationSetting < ApplicationRecord
     belongs_to :user, touch: true
 
-    validates :email_digest_periodic, inclusion: { in: [true, false] }
+    validates :email_badge_notifications, exclusion: [nil]
```
I'm still not super convinced by the need for these validations, as we ensure database integrity via non-null constraints and the UI should never generate a `nil` here. But as you said, we can decide on this later, it's ok to leave it in for now.

takmar replied:
> the UI should never generate a nil here.

When I read this, I have finally understood what I felt something wrong.

I was guessing these sending nil tests below are for a case that a user sends nil deliberately by rewriting HTML or doing something like that.
https://github.com/forem/forem/blob/1253d90114c0c795b36de31d10ab2476b716125d/spec/requests/user/user_notification_settings_spec.rb#L40-L44
https://github.com/forem/forem/blob/1253d90114c0c795b36de31d10ab2476b716125d/spec/requests/user/user_notification_settings_spec.rb#L85-L91

As you said, a user never sends nil usually, so I think it would be better to remove these tests.
And do so, we could also remove all the validations on the model.

What do you think?

citizen428 replied:
Ah, I think this is where we previously didn't quite manage to get on the same page 😃 I always approached this from the point of our UI where this is a checkbox:

![image](https://user-images.githubusercontent.com/47985/132794120-e03e4013-9b9c-49d8-9abc-f525a14acae3.png)

The way Rails handles these is via a hidden field with a value of "0", so an unchecked box also won't result in `nil`. 

So yes, I think we can remove these specs and the newly added validations, if someone really manages to send a `nil` here I'd be ok with the exception from the DB bubbling up because it would truly be an exceptional circumstance.

takmar replied:
>  I think we can remove these specs and the newly added validations

I removed the specs and the validations I added.
(I think email_digest_periodic validation is also unnecessary.)

> if someone really manages to send a nil here I'd be ok with the exception from the DB bubbling up because it would truly be an exceptional circumstance.

I totally agree with you 👍

msarit reviewed (approved):
Great work @takmar!
